SSZ-QL: calculate generalized indices for elements

fernantho opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
5. A changelog entry is required for user facing issues.
-->

**What type of PR is this?**
Feature

**Which issues(s) does this PR fix?**
Partially https://github.com/OffchainLabs/prysm/issues/15598

**What does this PR do? Why is it needed?**
This PR develop a way to calculate the Generalized Indices of a given path within a SSZ Object. To do so, it follows [Consensus Spec's Merkle proofs](https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md#ssz-object-to-index).

A conversion from PathElement to Generalized Index is necessary to work with fastssz proofs library. 

The code implemented walks the path within the SSZInfo struct in a consensus layer spec way. We have to take into account that the "path" input is slightly different:
```python
def get_generalized_index(typ: SSZType, *path: PyUnion[int, SSZVariableName]) -> GeneralizedIndex:
    """
    Converts a path (eg. `[7, "foo", 3]` for `x[7].foo[3]`, `[12, "bar", "__len__"]` for
    `len(x[12].bar)`) into the generalized index representing its position in the Merkle tree.
    """
```
```golang
// GetGeneralizedIndexFromPath calculates the generalized index for a given path.
// To calculate the generalized index, two inputs are needed:
// 1. The sszInfo of the root info, to be able to navigate the SSZ structure
// 2. The path to the field (e.g., "field_a.field_b[3].field_c")
// It walks the path step by step, updating the generalized index at each step.
func GetGeneralizedIndexFromPath(info *sszInfo, path []PathElement) (uint64, error) {
```
The pythonic version expects a Path input `[FieldA,"FieldB",3]` while the Go version expects `field_a.field_b[3]`.  

**Other notes for review**
- At the beginning, I considered implementing an implementation using recursion but this approach was discarded because of the inputs, as we already have the PathElement array, we can just loop this array.
- Input format should be snake case as the ssz generated type names are in snake case.
- Generalized Indices computed in tests were gotten from a Python script that relies on the Consensus Layer spec: https://github.com/fernantho/generalized-indices-ground-truth/blob/main/generalized_indices.ipynb
- Multidimensional arrays GI will be handled at a later iteration.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

fernantho committed:
added tests for calculating generalized indices

fernantho committed:
added first version of GI calculation walking the specified path with no recursion. Extended test coverage for bitlist and bitvectors.
vectors need more testing

fernantho committed:
refactored code. Detached PathElement processing, currently done at the beginning. Swap to regex to gain flexibility.

fernantho committed:
added an updateRoot function with the GI formula. more refactoring

fernantho committed:
added changelog

fernantho committed:
replaced TODO tag

fernantho committed:
udpated some comments

fernantho committed:
simplified code - removed duplicated code in processingLengthField function

fernantho committed:
Merge branch 'develop' into feat/ssz-ql-parse-path-to-generalized-index

fernantho committed:
run gazelle

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,358 @@
+// Helpers for input processing
+
+// processPathElement processes a path element string and returns an Element struct
+func processPathElement(elementStr string) (Element, error) {
```
I think these path-related processing can be done at `path.go`, so `GetGeneralizedIndexFromPath` won't process the path again. So `Element` in this file can be merged with `PathElement` (I'm not sure I've understood the PR description correctly but I think so)

One other concern for `elementStr` is that it's the result of `ParsePath`, which means the `index` information will probably be parsed. For example, if a path is like `"bitlist_field[1]"`, then `PathElement` would look like:

```go
&query.PathElement{
  Name: "bitlist_field",
  Index: &1 // pointer to uint64
}
```

fernantho replied:
Totally agree with this!

The key is to have a data type that covers all different current covered paths and leave open its extension for future ones, accessing multidimensional arrays, "filtering", "range" and/or "custom anchors" [we mentioned in our project proposal](https://hackmd.io/@junsong/HkQ9XBEHel).

I'll do the merge of this `Element` into the path.go.

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,358 @@
+		}
+		return (bitlistInfo.Limit() + 255) / bitsPerChunk, nil // Bits are packed into 256-bit chunks
+	case Bitvector:
+		vectorInfo, err := info.BitvectorInfo()
```
```suggestion
		bitvectorInfo, err := info.BitvectorInfo()
```

Just a nit!

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,358 @@
+			return 0, err
+		}
+		elemLength := itemLengthFromInfo(elementInfo)
+		return (vectorInfo.Length()*uint64(elemLength) + 31) / bytesPerChunk, nil
```
```suggestion
		return (vectorInfo.Length()*elemLength + 31) / bytesPerChunk, nil
```

Other nit, we don't need type cast here as it is already `uint64`. Same comment for `List` case above.

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,358 @@
+				return 0, fmt.Errorf("len() is only supported for List and Bitlist types, got %s", fieldSsz.sszType)
+			}
+			currentInfo = &sszInfo{sszType: UintN, fixedSize: 8}
+			root = updateRoot(root, 1, 2, 1)
```
```suggestion
			root = root*2 + 1
```

I think we can simply put the calculation here

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,358 @@
+			default:
+				return 0, fmt.Errorf("indexing not supported for type %s", fieldSsz.sszType)
+			}
+			continue
```
```suggestion
```
We might remove `continue` here

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,358 @@
+					multiplier = nextPowerOfTwo(innerChunkCount)
+					offset = *idx
+				}
+				root = updateRoot(root, 1, multiplier, offset)
```
Can we use `innerChunkCount` instead of `multiplier` here, as `List` and `Vector` would probably share same logic of calculating root index?

fernantho committed:
merging all input path processing into path.go

fernantho committed:
reviewed Jun's feedback

fernantho committed:
removed unnecessary idx pointer var + fixed error with length data type (uint64 instead of uint8)

fernantho committed:
refactored path.go after merging path elements from generalized_indices.go

fernantho committed:
re-computed GIs for tests as VariableTestContainer added a new field.

fernantho committed:
added minor comment - rawPath MUST be snake case

removed extractFieldName func.

fernantho committed:
fixed vector GI calculation - updated tests GIs

fernantho committed:
removed updateRoot function in favor of inline code

fernantho committed:
path input data enforced to be snake case

fernantho committed:
added sanity checks for accessing outbound element indices - checked against vector.length/list.limit

fernantho committed:
Merge branch 'develop' into feat/ssz-ql-parse-path-to-generalized-index

fernantho committed:
fixed issues triggered after merging develop

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -3,20 +3,30 @@ package query
 func ParsePath(rawPath string) ([]PathElement, error) {
+	// Enforce snake_case for field names
+	snakeCasePath := ToSnakeCase(rawPath)
 	// We use dot notation, so we split the path by '.'.
```
```suggestion
```
We might delete this comment as it seems duplicative with the comments on this function.

syjn99 commented on encoding/ssz/query/path_test.go:
```diff
@@ -34,6 +34,11 @@ func TestParsePath(t *testing.T) {
+			name:    "negative index in array path",
+			path:    ".data.target.root[-1]",
+			wantErr: true,
+		},
```
Would you mind adding more tests here? This PR will add a lot of functionality for `ParsePath`, but the unit tests aren't expressive enough.

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -32,31 +42,85 @@ func ParsePath(rawPath string) ([]PathElement, error) {
+var matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+func ToSnakeCase(str string) string {
```
```suggestion
func toSnakeCase(str string) string {
```
I'm pretty sure that we don't have to expose this function.

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -32,31 +42,85 @@ func ParsePath(rawPath string) ([]PathElement, error) {
+}
+
+var matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
```
Just wondering: Some regexes are pre-defined, others are compiled at runtime. Is there any reason?

fernantho replied:
Great observation! I hadn’t considered the performance aspect, but you’re right.
For example, I originally pulled the snake-case conversion from [this StackOverflow answer](https://stackoverflow.com/a/56616250), which pre-defines regexes and reuses them, which is the proper approach.
In contrast, here I’m recompiling the regex on every call:
```golang
func extractArrayIndices(name string) ([]uint64, error) {
    re := regexp.MustCompile(`$begin:math:display$\\s*([^$end:math:display$]+)\s*\]`)
    ...
}
```
I’ll refactor this to pre-compile the regex at the package level for consistency and efficiency.

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,275 @@
+		if err != nil {
+			return 0, err
+		}
+		// For Lists with basic element types, multiple elements can be packed into 32-byte chunks
```
I think this comment can be deleted, as `itemLengthFromInfo` will handle this logic. Same for `Vector` case.

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,275 @@
+			}
+			// Length is a uint64 per SSZ spec
+			currentInfo = &SszInfo{sszType: Uint64}
+			// root = root * base_index * pow2ceil(chunk_count(container)) + fieldPos
```
```suggestion
```

This comment seems irrelevant, so I would say we can delete this

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,275 @@
+		}
+
+		// Update the generalized index to point to the specified field
+		// root = root * base_index * pow2ceil(chunk_count(container)) + fieldPos
```
```suggestion
```
Code is quite explanationary, so I believe this comment can also be deleted. Same for other lines in this file

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,275 @@
+			continue
+		}
+
+		if element.Index != nil {
```
We might prevent unnecessary indentation if we `continue` early here when `element.Index == nil`

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,275 @@
+
+				currentInfo = elem
+
+			case Bitlist:
```
```suggestion
			case Bitlist, Bitvector:
```

Can we share the same logic like above?

fernantho replied:
So far, I'll leave with one case per type for clarity. 
The only difference is the base index for the root calculation. We would need to add another condition to disambiguate between Bitvector and Bitlist.

syjn99 replied:
Ah yes, base index is different. Then this must be okay 😅

syjn99 commented on encoding/ssz/query/generalized_index_test.go:
```diff
@@ -0,0 +1,338 @@
+	}
+}
+
+func TestGetIndicesFromPath_VariableTestContainer(t *testing.T) {
```
Can we add much more "crazier" tests like `.variable_container_list[0].inner_1.field_list_uint64[1]`?

fernantho replied:
Sure!
added more test cases

fernantho committed:
Removed redundant comment

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
removed unreachable condition as `strings.Split` always return a slice with length >= 1

If s does not contain sep and sep is not empty, Split returns a slice of
length 1 whose only element is s.

fernantho committed:
added tests to cover edge cases + cleaned code (toLower is no longer needed in extractFieldName function

fernantho committed:
added Jun's feedback + more testing

fernantho committed:
postponed snake case conversion to do it on a per-element-basis. Added more testing focused mainly in snake case conversion

fernantho committed:
addressed several Jun's comments.

fernantho committed:
added sanity check to prevent length of a multi-dimensional array. added more tests with extended paths

fernantho committed:
Merge branch 'develop' into feat/ssz-ql-parse-path-to-generalized-index

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+
+const (
+	bytesPerChunk = 32
+	bitsPerChunk  = 256
```
We already have `bytesPerChunk` defined in `encoding/ssz/helpers.go` - can you export and use it? And move `bitsPerChunk` over there too.

fernantho replied:
Sure! I hadn’t realized that `bytesPerChunk` was already defined there. It makes sense to keep both of those constants together in the helpers, while leaving `listBaseIndex` in encoding/ssz/query/generalized_index.go, as its domain is only for generalized indices.

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+
+// GetGeneralizedIndexFromPath calculates the generalized index for a given path.
+// To calculate the generalized index, two inputs are needed:
+// 1. The sszInfo of the root info, to be able to navigate the SSZ structure
```
```suggestion
// 1. The sszInfo of the root object, to be able to navigate the SSZ structure
```

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+// It walks the path step by step, updating the generalized index at each step.
+func GetGeneralizedIndexFromPath(info *SszInfo, path []PathElement) (uint64, error) {
+	if info == nil {
+		return 0, errors.New("sszInfo is nil")
```
```suggestion
		return 0, errors.New("SszInfo is nil")
```

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+		return 0, errors.New("sszInfo is nil")
+	}
+
+	// If path element list is empty, no generalized index can be computed.
```
```suggestion
	// If path is empty, no generalized index can be computed.
```

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+
+		// If we descend to a basic type, the path cannot continue further
+		if isBasicType(currentInfo.sszType) {
+			return 0, fmt.Errorf("cannot descend into basic type %s for path element %q", currentInfo.sszType, pathElement.Name)
```
What's the point of using `%q` and not `%s` for the element's name? I see there are other instances of using `%q` later in the file

fernantho replied:
Not a good one. I'm changing the code to use `%s` instead of `%q`.

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+		}
+
+		// Check that we are in a container to access fields
+		if currentInfo.sszType != Container {
```
Isn't having only this check sufficient? Do we need the above `isBasicType` check?

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+// Helpers for input processing
+
+// Copied from fastssz
+// Modified to return uint64
```
I would remove this comment because once we move away from fastszz it will become stale

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+		if err != nil {
+			return 0, err
+		}
+		return uint64(len(containerInfo.order)), nil
```
I think using `containerInfo.fields` will be more clear. "The length of order" is a bit cryptic in my opinion

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+			// Length is a uint64 per SSZ spec
+			currentInfo = &SszInfo{sszType: Uint64}
+			root = root*2 + 1
+			continue
```
Is it possible to move this to a helper function?

fernantho replied:
Regarding all these comments, yes. I really hate a `switch` clause without helpers 😅
In my experience with Go, I’ve dealt with a few protoactor `Receive()` handlers for message processing, and the pattern I liked the most were those that push as much logic as possible into helpers outside of the `switch`. That way, the different cases you’re dealing with are crystal clear.
In this case, I didn’t use helpers because I thought the cases were already clear, but maybe they only felt clear to me because I had just written the code. I’ll export all this logic to helpers so that new readers can also benefit from that clarity.

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+			}
+			// root = root * base_index * pow2ceil(chunk_count(container)) + fieldPos
+			root = root*listBaseIndex*nextPowerOfTwo(innerChunkCount) + chunkPos
+			currentInfo = elem
```
Is it possible to move this to a helper function?

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+			}
+			root = root*nextPowerOfTwo(innerChunkCount) + chunkPos
+
+			currentInfo = elem
```
Is it possible to move this to a helper function?

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+			root = root*listBaseIndex*nextPowerOfTwo(innerChunkCount) + chunkPos
+
+			// Bits element is not further descendable; set to basic to guard further steps
+			currentInfo = &SszInfo{sszType: Boolean}
```
Is it possible to move this to a helper function?

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,274 @@
+			root = root*nextPowerOfTwo(innerChunkCount) + chunkPos
+
+			// Bits element is not further descendable; set to basic to guard further steps
+			currentInfo = &SszInfo{sszType: Boolean}
```
Is it possible to move this to a helper function?

fernantho committed:
placed constant bitsPerChunk in the right place. Exported BitsPerChunk and BytesPerChunk and updated code that use them

fernantho committed:
Update encoding/ssz/query/generalized_index.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

fernantho committed:
Update encoding/ssz/query/generalized_index.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

fernantho committed:
Update encoding/ssz/query/generalized_index.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

fernantho committed:
added helpers for computing GI of each data type

fernantho committed:
Merge branch 'develop' into feat/ssz-ql-parse-path-to-generalized-index

fernantho committed:
changed %q in favor of %s

syjn99 commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,318 @@
+	default:
+		return false
+	}
+}
```
We can add `isBasic()` method under `SSZType` like:

```go
func (t SSZType) isBasic() bool {
	return t == Uint8 || t == Uint16 || t == Uint32 || t == Uint64 || t == Boolean
}
```

so that this file won't include this function. This seems more natural as the logic belongs to the `SSZType`.

fernantho replied:
Leaving this for tomorrow. Yes, I think that is the correct domain for `isBasic()` function.
I think some other functions can be rearranged, these are some candidates:
- `getChunkCount`
- `getContainerFieldByName`
- `itemLengthFromInfo`
- `nextPowerOfTwo`

fernantho replied:
I checked those candidates to be re-located but just arranged them (`getChunkCount`, `itemLengthFromInfo`, `nextPowerOfTwo`) altogether as part of SSZ helpers in the same file.

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -3,23 +3,28 @@ package query
+// 4. Handles leading dots and validates path format.
 func ParsePath(rawPath string) ([]PathElement, error) {
-	// We use dot notation, so we split the path by '.'.
+	// Enforce snake_case for field names
```
I think this comment can be deleted, seems like a remainder from previous commits.

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -31,32 +36,89 @@ func ParsePath(rawPath string) ([]PathElement, error) {
+			indices, err := extractArrayIndices(processingField)
 			if err != nil {
-				return nil, fmt.Errorf("invalid index in path element %s: %w", elem, err)
+				return []PathElement{}, err
```
```suggestion
				return nil, err
```

Same for the one at below.

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -31,32 +36,89 @@ func ParsePath(rawPath string) ([]PathElement, error) {
+var matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+func toSnakeCase(str string) string {
```
This feature looks great, but IMO it feels like adding a correction logic for not following the convention would be an overkill in this phase. We might return an error like `"path not found"` for paths using `CamelCase` or something. What do you think?

fernantho replied:
I agree with your point here. 

Maybe a correction logic is too much as this stage, even a validation of the path format is not necessary. I would consider it in case it early cuts ssz-ql query execution, offloading the client. 
But, in the current logic, execution stops as soon as a certain field is not found in the CL object, before we start requesting big data objects like the beacon state. 

Deleted this `toSnakeCase` logic

fernantho committed:
Update encoding/ssz/query/path.go

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
removed the least restrictive condition isBasicType

fernantho committed:
replaced length of containerInfo.order for containerInfo.fields for clarity

fernantho committed:
removed outdated comment

fernantho committed:
removed toSnakeCase conversion.

syjn99 commented on encoding/ssz/query/path_test.go:
```diff
@@ -7,6 +7,8 @@ import (
 	"github.com/OffchainLabs/prysm/v6/testing/require"
 )
 
+var index = uint64(0)
```
Unused one here!

fernantho committed:
moved isBasicType func to its natural place, SSZType

syjn99 commented on encoding/ssz/query/generalized_index_test.go:
```diff
@@ -0,0 +1,374 @@
+		{
+			name:          "Basic field cannot descend",
+			path:          "value1.value1",
+			expectedIndex: 0,
```
I think if `expectedError` is `true`, we can omit `expectedIndex`. Same comments for this test file.

fernantho replied:
Yes. Having `expectedIndex` when an error is expected is harmless but reduces clarity. Removed them all

syjn99 reviewed (commented):
Overall good, my unofficial approval 👍

fernantho committed:
cosmetic refactor

- renamed itemLengthFromInfo to itemLength (same name is in spec).
- arranged all SSZ helpers.

fernantho committed:
cleaned tests

fernantho committed:
Merge branch 'develop' into feat/ssz-ql-parse-path-to-generalized-index

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,306 @@
+	}
+
+	// Starting from the root generalized index
+	root := uint64(1)
```
Can you rename it to something like `currentIndex`? It's a bit odd to call it `root` since it's not a root of anything

fernantho replied:
Sure! 
I got the inspiration for the `root` name from spec:
```python
def get_generalized_index(typ: SSZType, *path: PyUnion[int, SSZVariableName]) -> GeneralizedIndex:
    """
    Converts a path (eg. `[7, "foo", 3]` for `x[7].foo[3]`, `[12, "bar", "__len__"]` for
    `len(x[12].bar)`) into the generalized index representing its position in the Merkle tree.
    """
    root = GeneralizedIndex(1)
   (...)
```
But I do not like it because I do not associate it to an index.

rkapka commented on encoding/ssz/query/path.go:
```diff
@@ -32,31 +36,78 @@ func ParsePath(rawPath string) ([]PathElement, error) {
+// e.g. "array[0][1]" -> []uint64{0, 1}. Errors if none are found or if any index is invalid.
+func extractArrayIndices(name string) ([]uint64, error) {
+	// Match all bracketed content, then we'll parse as unsigned to catch negatives explicitly
+	re := arrayIndexRegex
```
What's the purpose of assigning to a new variable instead of using `arrayIndexRegex` directly?

fernantho replied:
No purpose at all. I'll remove it.

rkapka commented on encoding/ssz/query/path.go:
```diff
@@ -32,31 +36,78 @@ func ParsePath(rawPath string) ([]PathElement, error) {
+		idx, err := strconv.ParseUint(raw, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("invalid array index: %w", err)
+		}
```
Looking at the documentation for `ParseUint` it seems that it doesn't allow negative numbers, so the `HasPrefix` check is unnecessary
> // ParseUint is like [ParseInt] but for unsigned numbers.
//
// **_A sign prefix is not permitted._**

fernantho replied:
Thanks. Totally unnecessary check.

rkapka commented on encoding/ssz/query/path.go:
```diff
@@ -32,31 +36,78 @@ func ParsePath(rawPath string) ([]PathElement, error) {
-			if len(parts) != 2 {
-				return nil, fmt.Errorf("invalid index notation in path element %s", elem)
-			}
+		re := regexp.MustCompile(`^\s*len\s*\(\s*([^)]+?)\s*\)\s*$`)
```
How about extracting this regex to a package-level variable, just like you did with `arrayIndexRegex`?

fernantho replied:
I forgot to extract this one. It's now at the package-level variable.

rkapka commented on encoding/ssz/query/path.go:
```diff
@@ -32,31 +36,78 @@ func ParsePath(rawPath string) ([]PathElement, error) {
-			}
+		re := regexp.MustCompile(`^\s*len\s*\(\s*([^)]+?)\s*\)\s*$`)
+		matches := re.FindStringSubmatch(processingField)
+		if len(matches) == 2 {
```
It would be nice to add an explanation why `2` is expected

fernantho replied:
Added explanation. I'm now considering regular expressions an overkill, when I added them I was considering input validation and correction.

rkapka commented on encoding/ssz/query/path.go:
```diff
@@ -32,31 +36,78 @@ func ParsePath(rawPath string) ([]PathElement, error) {
+			// only a single index is supported per PathElement, e.g., "transactions[0]" is valid
+			// while "transactions[0][0]" is rejected explicitly.
+			if len(indices) != 1 {
+				return []PathElement{}, fmt.Errorf("multiple indices not supported in token %q", processingField)
```
```suggestion
				return []PathElement{}, fmt.Errorf("multiple indices not supported in token %s", processingField)
```

rkapka commented on encoding/ssz/query/path_test.go:
```diff
@@ -34,14 +34,34 @@ func TestParsePath(t *testing.T) {
 			},
 			wantErr: false,
 		},
+		{
```
Can you add even more test cases? The ones I can think of:
- leading double dot --> error
- trailing dot --> error
- `len(data)` --> error
- `len(data.target.root)` --> ok
- `len(data.target.root).foo` --> error
- `data.target.len(root)` --> error

The easiest way to get a bunch of test cases it to pass the regex to an AI and ask it to generate them. I expect it will overdo this, but some cases can be useful.

fernantho replied:
Adding them, but I see we do not follow the same convention for requests:

> * `len(data.target.root)` --> error
> * `data.target.len(root)` --> ok

As we firstly split the raw input by `.`, having this input `len(data.target.root)` would result in:
- `len(data`
- `target`
- `root)`

leading to a wrong outcome.

On the contrary, this would succeed `data.target.len(root)`
- `data`
- `target`
- `len(root)`

fernantho replied:
<img width="356" height="231" alt="image" src="https://github.com/user-attachments/assets/e739582c-f62a-4135-b0e5-cfa02fef588a" />

<img width="373" height="232" alt="image" src="https://github.com/user-attachments/assets/882210d1-5f01-4f04-802d-074ddc1d9501" />

We must properly specify the input format for these queries.

fernantho replied:
> * `len(data)` --> error

I do not see an error here neither. Imagine we are querying `validators` field length in beacon state.
In this case the query would contain `len(validators)`

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,306 @@
+// Helpers for Generalized Index calculation per type
+
+// calculateLengthGeneralizedIndex calculates the generalized index for a length field.
+// note: length fields are only valid for List and Bitlist types. Multi-dimensional arrays are not supported.
```
Shouldn't this be supported for `Vector` and `Bitvector` too?

fernantho replied:
In relation to this, I also followed the spec algo:
```diff
    for p in path:
        # If we descend to a basic type, the path cannot continue further
        assert not issubclass(typ, BasicValue)
        if p == "__len__":
+            assert issubclass(typ, (List, ByteList))
            typ = uint64
            root = GeneralizedIndex(root * 2 + 1)
```
To my understanding, there is no length field for `Vector` and `Bitvector` as they have fixed size determined by their type.

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,306 @@
+
+// calculateLengthGeneralizedIndex calculates the generalized index for a length field.
+// note: length fields are only valid for List and Bitlist types. Multi-dimensional arrays are not supported.
+func calculateLengthGeneralizedIndex(fieldSsz *SszInfo, element PathElement, root uint64) (*SszInfo, uint64, error) {
```
Similarly to the comment above (changing `root` to `currentIndex`), can you rename the `root` param to something like `parentIndex` (or something else more suitable)?

fernantho replied:
Re-renamed them 😅 
changed all of them from `root` to `currentIndex`. But now I got to this comment and, for these functions, I like this `parentIndex` more than `currentIndex`.

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -0,0 +1,306 @@
+// note: length fields are only valid for List and Bitlist types. Multi-dimensional arrays are not supported.
+func calculateLengthGeneralizedIndex(fieldSsz *SszInfo, element PathElement, root uint64) (*SszInfo, uint64, error) {
+	if element.Index != nil {
+		return nil, 0, fmt.Errorf("len() is not supported for indexed elements (multi-dimensional arrays)")
```
```suggestion
		return nil, 0, fmt.Errorf("len() is not supported for multi-dimensional arrays")
```
Would that message be equivalent? Because I find the original one a little hard to understand

fernantho replied:
The original message is very shady. Committed your suggestion!

fernantho committed:
renamed "root" to "index"

fernantho committed:
removed unnecessary check for negative integers. Replaced %q for %s.

fernantho committed:
refactored regex variables and prevented re-assignation

fernantho committed:
added length regex explanation

fernantho committed:
added more testing for stressing regex for path processing

fernantho committed:
renamed currentIndex to parentIndex for clarity and documented the returns from calculate<Type>GeneralizedIndex functions

fernantho committed:
Update encoding/ssz/query/generalized_index.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

rkapka committed:
Merge branch 'develop' into feat/ssz-ql-parse-path-to-generalized-index

fernantho committed:
run gazelle

fernantho committed:
fixed never asserted error. Updated error message
